Add page sitemap provider for new Sulu 3.0 PageBundle

martinlagler opened the pull request:
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | no
| BC breaks? | yes
| Related Issues | #7672
| License | MIT

#### What's in this PR?

Update PageSiteMapProvider for the Sulu 3.0 structure.

Prokyonn commented on packages/page/src/Infrastructure/Sulu/Sitemap/PagesSitemapProvider.php:
```diff
@@ -0,0 +1,275 @@
+    /**
+     * @var EntityRepository<Route>
+     */
+    protected NestedTreeRepository $routeRepository;
```
RouteRepository can be removed, I think it is not used anywhere

Prokyonn commented on packages/page/src/Infrastructure/Sulu/Sitemap/PagesSitemapProvider.php:
```diff
@@ -0,0 +1,275 @@
+    /**
+     * @var EntityRepository<PageInterface>
+     */
+    protected NestedTreeRepository $entityRepository;
```
As long as you do not need any methods from the `NestedTreeRepository` I would type it with `EntityRepository`

Prokyonn commented on packages/page/src/Infrastructure/Sulu/Sitemap/PagesSitemapProvider.php:
```diff
@@ -0,0 +1,275 @@
+    public function build($page, $scheme, $host)
+    {
+        $portalInformations = $this->webspaceManager->findPortalInformationsByHostIncludingSubdomains(
+            $host, $this->environment
```
```suggestion
            $host, 
            $this->environment
```

Prokyonn commented on packages/page/src/Infrastructure/Sulu/Sitemap/PagesSitemapProvider.php:
```diff
@@ -0,0 +1,275 @@
+        $this->routeRepository = $routeRepository;
+    }
+
+    public function build($page, $scheme, $host)
```
Add types to the method parameters of the interface, also return type is missing

now is the time to do BC breaks :D

Prokyonn commented on packages/page/src/Infrastructure/Sulu/Sitemap/PagesSitemapProvider.php:
```diff
@@ -0,0 +1,275 @@
+            }
+
+            $locale = $portalInformation->getLocalization()->getLocale();
+            $webspaceKey = $this->webspaceManager->findPortalByKey($portalInformation->getPortalKey())->getWebspace()->getKey();
```
I think the `PortalInformation` class should already have a "getWebspace" method 🤔

Prokyonn commented on packages/page/src/Infrastructure/Sulu/Sitemap/PagesSitemapProvider.php:
```diff
@@ -0,0 +1,275 @@
+
+        $query = $queryBuilder->getQuery();
+
+        return $query->getArrayResult();
```
```suggestion
        return $query->toIterable();
```

Prokyonn commented on packages/page/src/Infrastructure/Sulu/Sitemap/PagesSitemapProvider.php:
```diff
@@ -0,0 +1,275 @@
+        return $query->getArrayResult();
+    }
+
+    private function generateSitemapUrl(
```
phpdoc for the arrays and. return type missing

alexander-schranz commented on src/Sulu/Bundle/WebsiteBundle/Sitemap/SitemapUrl.php:
```diff
@@ -48,7 +48,7 @@ class SitemapUrl
      * Datetime of last modification.
      *
-     * @var \DateTime
+     * @var \DateTimeImmutable|null
```
@martinlagler use `\DateTimeInterface` currently

alexander-schranz commented on packages/page/src/Infrastructure/Sulu/Sitemap/PagesSitemapProvider.php:
```diff
@@ -0,0 +1,295 @@
+use Sulu\Page\Domain\Model\PageInterface;
+
+/**
+ * @phpstan-type Page array{
```
```suggestion
/**
 * @internal your code should not create direct dependencies on this implementation
 *           projects can create there own sitemap providers or use symfony 
 *           dependency injection container to override services
 *
 * @phpstan-type Page array{
```

alexander-schranz commented on packages/page/tests/Functional/Integration/PageControllerTest.php:
```diff
@@ -180,7 +180,7 @@ public function testPostPublish(): string
     public function testVersionListAfterPublish(string $id): string
     {
-        $this->client->request('GET', '/admin/api/pages/' . $id . '/versions?page=1&locale=en&webspace=sulu-io&fields=title,version,changer,id');
+        $this->client->request('GET', 'http://sulu.io/admin/api/pages/' . $id . '/versions?page=1&locale=en&webspace=sulu-io&fields=title,version,changer,id');
```
the `/admin` urls don't need to be changed only the website urls

martinlagler committed:
Update page sitemap provider

alexander-schranz committed:
Enhance internal documentation for PagesSitemapProvider

Added internal documentation regarding dependency management for sitemap providers.

alexander-schranz committed:
Undo interface change

alexander-schranz committed:
Fix after undo interface change

alexander-schranz committed:
Undo undocumented Interface change

alexander-schranz committed:
Change parameter types in build method
